File: intellij/src/test/kotlin/com/rsunder10/formatincrementer/SequenceTest.kt
package com.rsunder10.formatincrementer

import org.junit.Assert.assertEquals
import org.junit.Test

class SequenceTest {

    private fun render(config: PatternConfig, count: Int) =
        (0 until count).map { Sequence.render(config, it) }

    @Test
    fun numeric() {
        val c = PatternConfig(type = PatternType.NUMERIC, start = 1, step = 2)
        assertEquals(listOf("1", "3", "5"), render(c, 3))
    }

    @Test
    fun padding() {
        val c = PatternConfig(type = PatternType.PADDING, start = 1, step = 1, width = 3)
        assertEquals(listOf("001", "002", "003"), render(c, 3))
    }

    @Test
    fun radixHexUppercase() {
        val c = PatternConfig(type = PatternType.RADIX, start = 10, step = 5, radix = 16, uppercase = true)
        assertEquals(listOf("A", "F", "14"), render(c, 3))
    }

    @Test
    fun alpha() {
        val c = PatternConfig(type = PatternType.ALPHA, start = 1, step = 1)
        assertEquals(listOf("a", "b", "c"), render(c, 3))
        val c27 = PatternConfig(type = PatternType.ALPHA, start = 27, step = 1)
        assertEquals("aa", Sequence.render(c27, 0))
    }

    @Test
    fun roman() {
        val c = PatternConfig(type = PatternType.ROMAN, start = 1, step = 1, uppercase = true)
        assertEquals(listOf("I", "II", "III"), render(c, 3))
    }

    @Test
    fun dates() {
        val c = PatternConfig(
            type = PatternType.DATES,
            startDate = "2024-01-01T00:00:00",
            stepAmount = 1,
            stepUnit = StepUnit.DAYS,
            dateFormat = "yyyy-MM-dd",
        )
        assertEquals(listOf("2024-01-01", "2024-01-02", "2024-01-03"), render(c, 3))
    }

    @Test
    fun template() {
        val c = PatternConfig(type = PatternType.TEMPLATE, start = 1, step = 1, template = "row-{n:03}")
        assertEquals(listOf("row-001", "row-002"), render(c, 2))
        val hex = PatternConfig(type = PatternType.TEMPLATE, start = 10, step = 1, template = "0x{n:hex}")
        assertEquals("0xa", Sequence.render(hex, 0))
    }
}

File: vscode/src/test/sequence.test.ts
import * as assert from "assert";
import { PatternConfig, DEFAULT_CONFIG, render } from "../sequence";

function make(overrides: Partial<PatternConfig>): PatternConfig {
  return { ...DEFAULT_CONFIG, ...overrides };
}

function seq(config: PatternConfig, count: number): string[] {
  const out: string[] = [];
  for (let i = 0; i < count; i++) {
    out.push(render(config, i));
  }
  return out;
}

const tests: Array<[string, () => void]> = [
  [
    "numeric",
    () =>
      assert.deepStrictEqual(
        seq(make({ type: "numeric", start: 1, step: 2 }), 3),
        ["1", "3", "5"]
      ),
  ],
  [
    "padding",
    () =>
      assert.deepStrictEqual(
        seq(make({ type: "padding", start: 1, step: 1, width: 3 }), 3),
        ["001", "002", "003"]
      ),
  ],
  [
    "radix hex uppercase",
    () =>
      assert.deepStrictEqual(
        seq(make({ type: "radix", start: 10, step: 5, radix: 16, uppercase: true }), 3),
        ["A", "F", "14"]
      ),
  ],
  [
    "alpha",
    () => {
      assert.deepStrictEqual(seq(make({ type: "alpha", start: 1, step: 1 }), 3), [
        "a",
        "b",
        "c",
      ]);
      assert.strictEqual(render(make({ type: "alpha", start: 27, step: 1 }), 0), "aa");
    },
  ],
  [
    "roman",
    () =>
      assert.deepStrictEqual(
        seq(make({ type: "roman", start: 1, step: 1, uppercase: true }), 3),
        ["I", "II", "III"]
      ),
  ],
  [
    "dates",
    () =>
      assert.deepStrictEqual(
        seq(
          make({
            type: "dates",
            startDate: "2024-01-01T00:00:00",
            stepAmount: 1,
            stepUnit: "days",
            dateFormat: "yyyy-MM-dd",
          }),
          3
        ),
        ["2024-01-01", "2024-01-02", "2024-01-03"]
      ),
  ],
  [
    "template",
    () => {
      assert.deepStrictEqual(
        seq(make({ type: "template", start: 1, step: 1, template: "row-{n:03}" }), 2),
        ["row-001", "row-002"]
      );
      assert.strictEqual(
        render(make({ type: "template", start: 10, step: 1, template: "0x{n:hex}" }), 0),
        "0xa"
      );
    },
  ],
];

let failed = 0;
for (const [name, fn] of tests) {
  try {
    fn();
    console.log(`ok - ${name}`);
  } catch (e) {
    failed++;
    console.error(`FAIL - ${name}`);
    console.error(e);
  }
}

if (failed > 0) {
  console.error(`${failed} test(s) failed`);
  process.exit(1);
} else {
  console.log(`All ${tests.length} tests passed`);
}
